sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -368,6 +368,15 @@ abstract class Expression extends TreeNode[Expression] {
     } else {
       ""
     }
+
+
+  def supportWholeStageCodegen(): Boolean = {
+    this.getClass.getDeclaredMethods.exists(_.getName == "nullSafeEval") &&
```

Comment:
I think this is a working but hacky solution. I'd like to re-implement these `CodegenFallback` expressions using `Invoke` or `StaticInvoke`, and eventually remove `CodegenFallback`

Reply:
The expressions which don't have `nullSafeEval()`  method, still need fallback to evaluate `eval()` method and not support code-gen.

Reply:
I said this works. `nullsafeEval` is not designed for this purpose and I think it's fair to say this solution is hacky. You happened to find a way to use `nullsafeEval` for codegen. A cleaner solution is to use `Invoke` or `StaticInvoke`, to explicitly define the java method with input values for evaluating the expressions.